pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+)
+
+const (
+	// APIGroupAll matches all api groups.
+	APIGroupAll = "*"
+	// ResourceAll matches all api resources.
+	ResourceAll = "*"
+	// NonResourceAll matches all api non-resource paths.
+	NonResourceAll = "*"
+	// VerbAll matches all request verbs.
+	VerbAll = "*"
+
+	// GroupKind represents "group" subject kind
+	GroupKind = "Group"
+	// ServiceAccountKind represents "service-account" subject kind
+	ServiceAccountKind = "ServiceAccount"
+	// UserKind represents "user" subject kind
+	UserKind = "User"
+)
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition;
+// "ConditionFalse" means a resource is not in the condition; "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	FlowSchemaConditionTrue    FlowSchemaConditionStatus = "True"
+	FlowSchemaConditionFalse   FlowSchemaConditionStatus = "False"
+	FlowSchemaConditionUnknown FlowSchemaConditionStatus = "Unknown"
+
+	PriorityLevelConfigurationConditionTrue    PriorityLevelConfigurationConditionStatus = "True"
+	PriorityLevelConfigurationConditionFalse   PriorityLevelConfigurationConditionStatus = "False"
+	PriorityLevelConfigurationConditionUnknown PriorityLevelConfigurationConditionStatus = "Unknown"
+)
+
+// These are valid flow-dingtinguisher methods.
+const (
+	FlowDistinguisherMethodNone        = "None"
```

Comment:
Note that this is _not_ the zero value of its type.

Reply:
It's OK if this is a default, it's also OK to require the user to explicitly choose (no default, "" not allowed).

Reply:
Do we have a compelling use for "none"? It is dangerous, it defeats the purpose of the shuffle sharding.

Reply:
Ah, I guess you can use "none" if the PL has only 1 queue? But it also doesn't matter in that case.

Reply:
I _thought_ we earlier agreed that the zero value of `FlowDistinguisherMethod` would prescribe the degenerate method (the distinguisher is always `""`).  When there is only one queue, there is no point in prescribing a non-degenerate FlowDistinguisherMethod.  Because of the indirection from flow to queue it is true that a non-degenerate FlowDistinguisherMethod is not technically a problem.  But it is pointless.